test/server/drain_manager_impl_test.cc

```diff
@@ -151,8 +159,398 @@ TEST_P(DrainManagerImplTest, DrainDeadlineProbability) {
+        EXPECT_THAT(delay.count(), AllOf(Ge(i * step - 1), Le(i * step + 1)));
+      }));
+
+      cb_handles[i] = drain_manager.addOnDrainCloseCb(cb.AsStdFunction());
+    }
+
+    drain_manager.startDrainSequence([] {});
+  }
+
+  EXPECT_TRUE(drain_manager.draining());
+}
+
+// Validate the expected behavior when a drain-close callback is registered after draining has begun
+// with a Gradual drain strategy (should be called with delay between 0 and maximum)
+TEST_F(DrainManagerImplTest, RegisterCallbackAfterDrainBeginGradualStrategy) {
+  ON_CALL(server_.options_, drainStrategy()).WillByDefault(Return(Server::DrainStrategy::Gradual));
+  ON_CALL(server_.options_, drainTime()).WillByDefault(Return(std::chrono::seconds(1)));
+
+  DrainManagerImpl drain_manager(server_, envoy::config::listener::v3::Listener::DEFAULT,
+                                 server_.dispatcher());
+
+  testing::MockFunction<void(std::chrono::milliseconds)> cb_before_drain;
+  testing::MockFunction<void(std::chrono::milliseconds)> cb_after_drain;
+
+  EXPECT_CALL(cb_before_drain, Call(_));
+  EXPECT_CALL(cb_after_drain, Call(_)).WillOnce(Invoke([](std::chrono::milliseconds delay) {
+    // Validate that callbacks after the drain sequence has started are called with a random value
+    // between 0 (immediate) and the max drain window.
+    EXPECT_THAT(delay.count(), Ge(0));
+    EXPECT_THAT(delay.count(), Le(1000));
+  }));
+
+  auto before_handle = drain_manager.addOnDrainCloseCb(cb_before_drain.AsStdFunction());
+  drain_manager.startDrainSequence([] {});
+  auto after_handle = drain_manager.addOnDrainCloseCb(cb_after_drain.AsStdFunction());
+  EXPECT_EQ(after_handle, nullptr);
+}
+
+// Validate the expected behavior when a drain-close callback is registered after draining has begun
+// with an Immediate drain strategy (should be called with 0 delay)
+TEST_F(DrainManagerImplTest, RegisterCallbackAfterDrainBeginImmediateStrategy) {
+  ON_CALL(server_.options_, drainStrategy()).WillByDefault(Return(Server::DrainStrategy::Gradual));
+  ON_CALL(server_.options_, drainTime()).WillByDefault(Return(std::chrono::seconds(1)));
+
+  DrainManagerImpl drain_manager(server_, envoy::config::listener::v3::Listener::DEFAULT,
+                                 server_.dispatcher());
+
+  testing::MockFunction<void(std::chrono::milliseconds)> cb_before_drain;
+  testing::MockFunction<void(std::chrono::milliseconds)> cb_after_drain;
+
+  EXPECT_CALL(cb_before_drain, Call(_));
+  EXPECT_CALL(cb_after_drain, Call(_)).WillOnce(Invoke([](std::chrono::milliseconds delay) {
+    EXPECT_EQ(delay.count(), 0);
+  }));
+
+  auto before_handle = drain_manager.addOnDrainCloseCb(cb_before_drain.AsStdFunction());
+  drain_manager.startDrainSequence([] {});
+  auto after_handle = drain_manager.addOnDrainCloseCb(cb_after_drain.AsStdFunction());
+  EXPECT_EQ(after_handle, nullptr);
+}
+
+// Destruction doesn't trigger draining, so it should be for a child's parent to be cleaned up
```

Comment:
nit: should be safe for a child's parent...?

Reply:
re-worded. :-)